File: lw6/Motorina_Eugenia/lab6/lab6/Sorter.cpp
#include "stdafx.h"
#include "ThreadPool.h"
#include "Sorter.h"
#include <math.h>

CSorter::CSorter()
{
}


CSorter::~CSorter()
{
}

void CSorter::Sort(size_t begin, size_t end)
{
	if (begin != end)
	{
		if (end - begin != 1)
		{
			size_t half = (end + begin) / 2;
			Sort(begin, half);
			Sort(half + 1, end);
			std::vector<int> buf;
			buf.resize(m_array.size());
			size_t xl = begin;
			size_t xr = half + 1;
			size_t cur = 0;
			while (end - begin + 1 != cur)
			{
				if (xl > half)
				{
					buf[cur++] = m_array[xr++];
				}
				else if (xr > end)
				{
					buf[cur++] = m_array[xl++];
				}
				else if (m_array[xl] > m_array[xr])
				{
					buf[cur++] = m_array[xr++];
				}
				else
				{
					buf[cur++] = m_array[xl++];
				}
			}
			for (int i = 0; i < cur; i++)
			{
				m_array[i + begin] = buf[i];
			}
		}
		else if (m_array[end] < m_array[begin])
		{
			std::swap(m_array[end], m_array[begin]);
		}
	}
}

void CSorter::StartThreads(size_t count)
{
	ThreadPool threadPool;
	auto size = m_array.size();
	auto block = size_t(std::ceil(double(size) / count));
	size_t currentPos = 0;
	for (size_t threadId = 0; threadId < count; ++threadId)
	{
		if (currentPos + block >= size)
		{
			threadPool.AddThread(std::thread(boost::bind(&CSorter::Sort, this, currentPos, size - 1)));
			break;
		}
		threadPool.AddThread(std::thread(boost::bind(&CSorter::Sort, this, currentPos, currentPos + block)));
		currentPos += block + 1;
	}
	threadPool.WaitThread();
}

void CSorter::MergeSort(std::vector<int> & array)
{
	m_array = array;
	Sort(0, m_array.size() - 1);
	array = m_array;
}

void CSorter::MergeSortWithThreads(std::vector<int>& array, size_t threadCount)
{
	m_array = array;
	StartThreads(threadCount);
	Sort(0, m_array.size() - 1);
	array = m_array;
}

File: lw7/MotorinaEvgeniya/lw7/lw7/Sorter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lw7
{
    class Sorter
    {
        public List<int> m_array = new List<int>();

        public void MergeSort(List<int> array)
        {
            m_array = array;
            Sort(0, m_array.Count() - 1);
            array = m_array;
        }

        public void SortWithThreads(List<int> array, int threadCount)
        {
            m_array = array;
            StartThreads(threadCount);
            Sort(0, m_array.Count() - 1);
            array = m_array;

        }

	void Sort(int b, int e)
        {
            int begin = b;
            int end = e;

            if (begin != end)
            {
                if (end - begin != 1)
                {
                    int half = (end + begin) / 2;
                    Sort(begin, half);
                    Sort(half + 1, end);
                    int[] buf = new int[m_array.Count()];
                    int xl = begin;
                    int xr = half + 1;
                    int cur = 0;
                    while (end - begin + 1 != cur)
                    {
                        if (xl > half)
                        {
                            buf[cur++] = m_array[xr++];
                        }
                        else if (xr > end)
                        {
                            buf[cur++] = m_array[xl++];
                        }
                        else if (m_array[xl] > m_array[xr])
                        {
                            buf[cur++] = m_array[xr++];
                        }
                        else
                        {
                            buf[cur++] = m_array[xl++];
                        }
                    }
                    for (int i = 0; i < cur; i++)
                    {
                        m_array[i + begin] = buf[i];
                    }
                }
                else if (m_array[end] < m_array[begin])
                {
                    int temp = m_array[end];
                    m_array[end] = m_array[begin];
                    m_array[begin] = temp;
                }
            }
        }

    void StartThreads(int count)
        {
            ThreadPool threadPool = new ThreadPool();
            int size = m_array.Count();
            int block = (int)Math.Floor((double)size / count);
            int currentPos = 0;
            for (int threadId = 0; threadId < count && currentPos < size; ++threadId)
            {
                int end = Math.Min(size - 1, currentPos + block);
                int begin = currentPos;
                threadPool.AddThread(Task.Factory.StartNew(() => Sort(begin, end)));
                currentPos += block + 1;

            }
            threadPool.WaitThread();
        }
    }
}
